ygen/enumgen.go

```diff
@@ -347,31 +347,81 @@ func (s *enumGenState) resolveEnumName(e *yang.Entry, compressPaths, noUnderscor
-		return definedName
+// resolveEnumName takes a yang.Entry and resolves its name into the type name
+// that will be used in the generated code. Whilst a leaf may only be used
+// in a single context (i.e., at its own path), resolveEnumName may be called
+// multiple times, and hence de-duplication of unique name generation is required.
+// If noUnderscores is set to true, then underscores are omitted from the
+// output name.
+// If the skipDedup argument is set to true, where a single enumeration is defined
+// once in the input YANG, but instantiated multiple times (e.g., a grouping is
+// used multiple times that contains an enumeration), then we do not attempt to
+// use a single output type in the generated code for such enumerations. This allows
+// the user to control whether this behaviour is useful to them -- for OpenConfig,
+// it tends to be due to the state/config split - which would otherwise result in
+// multiple enumerated types being produced. For other schemas, it can result in
+// somewhat difficult to understand enumerated types being produced - since the first
+// leaf that is processed will define the name of the enumeration.
+func (s *enumGenState) resolveEnumName(e *yang.Entry, compressPaths, noUnderscores, skipDedup bool) string {
+	// uniqueIdentifer is the unique identifier used to determine whether to
+	// define a new enum type for the input enum.
+	var uniqueIdentifer string
+	if skipDedup && !compressPaths {
+		// If not using compression and duplicating, then we use the
+		// entire path as the enum name, meaning every enum instance
+		// has its own definition. By using the entry's path as its
+		// identifier, we ensure this.
+		uniqueIdentifer = e.Path()
+	} else {
+		// In the other cases, de-duplication may happen, either
+		// through de-duping multiple usages or possibly through
+		// compression
+		uniqueIdentifer = enumIdentifier(e, compressPaths)
 	}
 
+	var compressName string
 	if compressPaths {
+		definingModName := genutil.ParentModulePrettyName(e.Node)
 		// If we compress paths then the name of this enum is of the form
 		// ModuleName_GrandParent_Leaf - we use GrandParent since Parent is
 		// State or Config so would not be unique. The proposed name is
 		// handed to genutil.MakeNameUnique to ensure that it does not clash with
 		// other defined names.
-		name := fmt.Sprintf("%s_%s_%s", yang.CamelCase(definingModName), yang.CamelCase(e.Parent.Parent.Name), yang.CamelCase(e.Name))
+		compressName = fmt.Sprintf("%s_%s_%s", yang.CamelCase(definingModName), yang.CamelCase(e.Parent.Parent.Name), yang.CamelCase(e.Name))
 		if noUnderscores {
-			name = strings.Replace(name, "_", "", -1)
+			compressName = strings.Replace(compressName, "_", "", -1)
 		}
-		uniqueName := genutil.MakeNameUnique(name, s.definedEnums)
-		s.uniqueEnumeratedLeafNames[identifierPath] = uniqueName
+
+		if skipDedup {
+			// If using compression and duplicating, then we add
```

Comment:
Clarified this wording here "and duplicating" seemed odd to me -- so I expanded it out to say that we're explicitly outputting a type per leaf.